# **Homework 2**
Name: Maddox Guthrie \
Id: 12184185

## 1.1 Exercise 1: Taylor Series Approximation
Use SymPy to compute the Taylor series of $f(x)$ about $0$ with the first $10$ non-zero
terms.

In [2]:
from sympy import symbols, exp, sin, series

x = symbols("x")

f = exp(x) * sin(2 * x)

x0 = 0
n = 11

f_series = series(f, x, x0, n)
t_series = f_series.removeO()
t_series

-779*x**10/907200 - 359*x**9/181440 + x**8/120 + 139*x**7/2520 + 11*x**6/180 - 19*x**5/60 - x**4 - x**3/3 + 2*x**2 + 2*x

Use the Taylor polynomial to approximate $f(0.4)$.

In [3]:
f_approx = t_series.subs({x:0.4})
f_approx

1.07016953578599

Compute the exact value of $f(0.4)$ using `math` package.

In [4]:
import math
f_actual = math.exp(0.4) * math.sin(2 * 0.4)
f_actual

1.0701695334073043

Compute the absolute error between the approximation and the actual value of the function
at $x = 0.4$.

In [5]:
err = abs(f_approx - f_actual)
err

2.37868547081632e-9

The approximation is very accurate. The absolute error is only $~2.38 \times 10^{-9}$, so the Taylor series matches $e^x sin(2x)$ to 8 or 9 significant figures at $x = 0.4$. This makes sense since $x = 0.4$ is close to the expansion point, so the series converges fast and the truncated terms are very small.

## 1.2 Exercise 2: Ordinary Differential Equations

### 1.2.1 Part A: Second-Order ODE with Forcing Term
Solve the non-homogeneous equation:
$$
y'' - 3y' + 2y = \sin(2t) - t^3 + e^t
$$

In [6]:
from sympy import symbols, Function, Eq, Derivative, dsolve

t = symbols("t")
y = symbols("y", cls=Function)

eq = Eq(Derivative(y(t), t, 2) - 3 * Derivative(y(t), t, 1) + 2 * y(t), sin(2 * t) - t ** 3 + exp(t))
solns = dsolve(eq, y(t))
solns

Eq(y(t), C2*exp(2*t) - t**3/2 - 9*t**2/4 - 21*t/4 + (C1 - t)*exp(t) - sin(2*t)/20 + 3*cos(2*t)/20 - 45/8)

### 1.2.2 Part B: Second-Order ODE with Initial Conditions
Solve the equation in Part A subject to:
$$
y(0) = 0, \quad y'(0) = 1
$$

In [7]:
solns = dsolve(eq, y(t), ics={y(t).subs(t, 0): 0, y(t).diff(t).subs(t, 0): 1})
solns

Eq(y(t), -t**3/2 - 9*t**2/4 - 21*t/4 + (18/5 - t)*exp(t) + 15*exp(2*t)/8 - sin(2*t)/20 + 3*cos(2*t)/20 - 45/8)

### 1.2.3 Part C: System of First-Order ODEs (Complex Eigenvalues)
Solve the system:
$$
\begin{cases}
x'(t) = x - 2y + sin(2t) \\
y'(t) = 2x + y + t^2
\end{cases}
$$

In [20]:
t = symbols("t")
x = symbols("x", cls=Function)
y = symbols("y", cls=Function)

eq1 = Eq(x(t).diff(t), x(t) - 2 * y(t) + sin(2 * t))
eq2 = Eq(y(t).diff(t), 2 * x(t) + y(t) + t**2)
eqs = [eq1, eq2]

solns = dsolve(eqs, [x(t), y(t)])
x_t = solns[0]
y_t = solns[1]

print("General Solution")
display(x_t, y_t)

General Solution


Eq(x(t), -C1*exp(t)*sin(2*t) - C2*exp(t)*cos(2*t) - 2*t**2*sin(2*t)**2/5 - 2*t**2*cos(2*t)**2/5 - 8*t*sin(2*t)**2/25 - 8*t*cos(2*t)**2/25 - 9*sin(2*t)**3/17 - 2*sin(2*t)**2*cos(2*t)/17 + 4*sin(2*t)**2/125 - 9*sin(2*t)*cos(2*t)**2/17 - 2*cos(2*t)**3/17 + 4*cos(2*t)**2/125)

Eq(y(t), C1*exp(t)*cos(2*t) - C2*exp(t)*sin(2*t) - t**2*sin(2*t)**2/5 - t**2*cos(2*t)**2/5 + 6*t*sin(2*t)**2/25 + 6*t*cos(2*t)**2/25 + 2*sin(2*t)**3/17 + 8*sin(2*t)**2*cos(2*t)/17 + 22*sin(2*t)**2/125 + 2*sin(2*t)*cos(2*t)**2/17 + 8*cos(2*t)**3/17 + 22*cos(2*t)**2/125)

### 1.2.4 Part D: System of ODEs with Initial Conditions
Solve the system in Part C subject to:
$$
x(0) = 1, \quad y(0) = 0
$$
Clearly distinguish between general and IVP solutions.

In [21]:
solns = dsolve(eqs, [x(t), y(t)], ics={x(0): 1, y(0): 0})
x_t = solns[0]
y_t = solns[1]

print("IVP Solution:")
display(x_t, y_t)

IVP Solution:


Eq(x(t), -2*t**2*sin(2*t)**2/5 - 2*t**2*cos(2*t)**2/5 - 8*t*sin(2*t)**2/25 - 8*t*cos(2*t)**2/25 + 1374*exp(t)*sin(2*t)/2125 + 2307*exp(t)*cos(2*t)/2125 - 9*sin(2*t)**3/17 - 2*sin(2*t)**2*cos(2*t)/17 + 4*sin(2*t)**2/125 - 9*sin(2*t)*cos(2*t)**2/17 - 2*cos(2*t)**3/17 + 4*cos(2*t)**2/125)

Eq(y(t), -t**2*sin(2*t)**2/5 - t**2*cos(2*t)**2/5 + 6*t*sin(2*t)**2/25 + 6*t*cos(2*t)**2/25 + 2307*exp(t)*sin(2*t)/2125 - 1374*exp(t)*cos(2*t)/2125 + 2*sin(2*t)**3/17 + 8*sin(2*t)**2*cos(2*t)/17 + 22*sin(2*t)**2/125 + 2*sin(2*t)*cos(2*t)**2/17 + 8*cos(2*t)**3/17 + 22*cos(2*t)**2/125)

## 1.3 Exercise 3: Critical Points and Hessian Classification
Consider the function
$$
f(x, y) = x^4 + y^4 - 4xy + 1
$$
Compute the first-order partial derivatives $f_x$ and $f_y$.

In [10]:
from sympy import symbols, diff

x, y = symbols("x y")
f = x**4 + y**4 - 4*x*y + 1

f_x = diff(f, x)
f_y = diff(f, y)

display(f_x, f_y)

4*x**3 - 4*y

-4*x + 4*y**3

Solve $\nabla f = 0$ to find all critical points.

In [11]:
from sympy import solve
critical_points = solve([f_x, f_y], [x, y])
critical_points

[(-1, -1),
 (0, 0),
 (1, 1),
 (-I, I),
 (I, -I),
 (sqrt(2)*(-1 - I)/2, sqrt(2)/2 - sqrt(2)*I/2),
 (sqrt(2)*(-1 + I)/2, sqrt(2)/2 + sqrt(2)*I/2),
 (sqrt(2)*(1 - I)/2, -sqrt(2)/2 - sqrt(2)*I/2),
 (sqrt(2)*(1 + I)/2, -sqrt(2)/2 + sqrt(2)*I/2)]

Compute the Hessian matrix of $f(x, y)$.

In [12]:
from sympy import hessian

h = hessian(f, [x, y])
h

Matrix([
[12*x**2,      -4],
[     -4, 12*y**2]])

Evaluate the Hessian at each critical point.

In [13]:
from sympy import det

display(det(h.subs({x: -1, y: -1})),
        det(h.subs({x: 0, y: 0})),
        det(h.subs({x: 1, y: 1})))

128

-16

128

Use the second derivative test to classify each point as:
- Local maximum
- Local minimum
- Saddle point

In [14]:
f_xx = diff(f_x, x)
display(f_xx.subs({x: -1, y: -1}),
        f_xx.subs({x: 0, y: 0}),
        f_xx.subs({x: 1, y: 1}))

12

0

12

$(-1,-1)$ Local minimum \
$(0,0)$ Saddle Point \
$(1,1)$ Local minimum